src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -8578,33 +8577,67 @@ static bool isCandidateUnique(ref MethodSymbol? method, MethodSymbol candidate)
+            {
+                // https://github.com/dotnet/roslyn/issues/55217: Support parameter
+                // and return types that are not valid generic type arguments.
+                return null;
+            }
+
+            bool hasByRefParameters = !parameterRefKinds.IsDefault && parameterRefKinds.Any(refKind => refKind != RefKind.None);
+
+            if (returnRefKind == RefKind.None && !hasByRefParameters)
             {
-                var wkDelegateType = returnTypeOpt.HasType ?
-                    WellKnownTypes.GetWellKnownFunctionDelegate(invokeArgumentCount: parameterTypes.Length) :
-                    WellKnownTypes.GetWellKnownActionDelegate(invokeArgumentCount: parameterTypes.Length);
+                var wkDelegateType = returnsVoid ?
+                    WellKnownTypes.GetWellKnownActionDelegate(invokeArgumentCount: parameterTypes.Length) :
+                    WellKnownTypes.GetWellKnownFunctionDelegate(invokeArgumentCount: parameterTypes.Length);
 
                 if (wkDelegateType != WellKnownType.Unknown)
                 {
                     var delegateType = Compilation.GetWellKnownType(wkDelegateType);
                     delegateType.AddUseSiteInfo(ref useSiteInfo);
-                    if (returnTypeOpt.HasType)
-                    {
-                        parameterTypes = parameterTypes.Add(returnTypeOpt);
-                    }
-                    if (parameterTypes.Length == 0)
+                    if (typeArguments.Length == 0)
                     {
                         return delegateType;
                     }
-                    if (checkConstraints(Compilation, Conversions, delegateType, parameterTypes))
+                    if (checkConstraints(Compilation, Conversions, delegateType, typeArguments))
                     {
-                        return delegateType.Construct(parameterTypes);
+                        return delegateType.Construct(typeArguments);
                     }
                 }
             }
 
-            return null;
+            var refKinds = (hasByRefParameters || returnRefKind != RefKind.None) ? RefKindVector.Create(parameterTypes.Length + (returnsVoid ? 0 : 1)) : default;
+            if (hasByRefParameters)
+            {
+                for (int i = 0; i < parameterRefKinds.Length; i++)
+                {
+                    refKinds[i] = parameterRefKinds[i];
+                }
+            }
+            if (returnRefKind != RefKind.None)
+            {
+                refKinds[parameterTypes.Length] = returnRefKind;
+            }
+
+            var synthesizedType = Compilation.AnonymousTypeManager.SynthesizeDelegate(parameterCount: parameterTypes.Length, refKinds, returnsVoid, generation: 0);
+            return synthesizedType.Construct(typeArguments);
+
+            static bool isValidTypeArgument(TypeSymbol? type)
+            {
+                return type is { } &&
+                    !type.IsPointerOrFunctionPointer() &&
```

Comment:
nit: Consider including a test with an array of pointer types (that's an example to illustrate the difference between `IsPointerOrFunctionPointer` and `IsUnsafe`)